### PR TITLE
support plan (New API PR 9)

### PR DESCRIPTION
**Pull request by kanej**
Add plan feature for the new api.

This involved:
1. Adding a new website as a package under `./packages/ui`
2. Adding serialization/deserialization for IgnitionModules
3. Update the plugin to write a plan folder under cache that interpolates the serialized version of the IgnitionModule being planned

The website is a React SPA using Vite. Everything is bundled into `index.html` so it can be opened locally without any CORs issues.

Serialization is done by flattening a module into a json document where each module/submodule is represented, but references to futures/modules are replaced by tokens. On deserialization, a first pass generates all of the relevant objects, and a second pass rewires them together.

## Preview

image

## Running locally

After a successful build a dev version of the plan website can be started with:

```sh
cd ./packages/ui
npm run dev
```

There is an example Module under `./examples/ComplexModule.js` that can be edited and used to update the website with:

```sh
npm run regenerate-deployment-example
```

Additionally, the `./examples/sample` project has been converted to use `defineModule` for testing:

```sh
cd ./examples/sample
hardhat plan ./ignition/LockModule.js
```

### CODE REVIEW

File: packages/core/src/new-api/stored-deployment-serializer.ts [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+      Object.entries(libraries).map(([key, token]) => [
+        key,
+        partialFutureLookup[token.futureId] as ContractFuture<string>,
+      ])
+    );
+  }
+
+  /**
+   * Oh you think you can defeat me typesystem. I don't acknowledge
+   * your _readonly_.
```
Comment (zoeyTM) [inline packages/core/src/new-api/stored-deployment-serializer.ts:330]: LOL brilliant

File: packages/ui/public/deployment.json [hunk 2]
```diff
@@ -0,0 +1,238 @@
+{
```
Comment (zoeyTM) [inline packages/ui/public/deployment.json:1]: is this meant to be included or ignored?

**Review by zoeyTM — APPROVED** [review]
I still have reservations about doing serialization/deserialization this way as opposed to simpler JSON, but this looks good. :+1:

File: packages/ui/public/deployment.json (hunk 2 above)
Comment (kanej) [reply packages/ui/public/deployment.json:1]: I had it there to help the development experience, but are you suggesting adding something like a "predev" script that runs `npm run regenerate-development-example` before running `npm run dev`?

File: packages/core/src/ui-helpers.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export { StoredDeploymentSerializer } from "./new-api/stored-deployment-serializer";
+export * from "./new-api/types/module";
+export { FutureType } from "./new-api/types/module";
```
Comment (alcuadrado) [inline packages/core/src/ui-helpers.ts:3]: Isn't this already exported in the line above?

```suggestion
```

File: packages/core/src/new-api/types/serialized-deployment.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+ * @beta
+ */
+export type SerializedSolidityParamsType = SerializedSolidityParamType[];
+
+/**
+ * The serialized version of the artifact.
+ *
+ * @beta
+ */
+export type SerializedArtifactType = any[];
```
Comment (alcuadrado) [inline packages/core/src/new-api/types/serialized-deployment.ts:22]: This can be the same type as the artifact, as they are already serializable.

**Review by alcuadrado — APPROVED** [review]

File: packages/core/src/ui-helpers.ts (hunk 3 above)
Comment (kanej) [reply packages/core/src/ui-helpers.ts]: Good catch. Thanks.

File: packages/ui/public/deployment.json (hunk 2 above)
Comment (kanej) [reply packages/ui/public/deployment.json]: I removed the `./public/deployment.json` from git and added the generation of it from the example to a prescript for `npm run dev`.

**Pull request merged**